File: sol/1937/1937.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        m = len(points)
        n = len(points[0])
        buff = [0] * n
        for i in range(1, m):
            opt = points[i - 1][0]
            buff[0] = opt + points[i][0]
            for j in range(1, n):
                opt = max(opt - 1, points[i - 1][j])
                buff[j] = opt + points[i][j]
            opt = 0
            points[i][n - 1] = buff[n - 1]
            for j in range(n - 2, -1, -1):
                opt = max(opt, points[i - 1][j + 1]) - 1
                points[i][j] = max(buff[j], opt + points[i][j])
        return max(points[m - 1])

File: sol/1937/1937.cpp
struct Solution {
    using ll = long long;
    long long maxPoints(vector<vector<int>>& points) {
        int m = points.size(), n = points[0].size();
        vector<ll> buff[2] = {vector<ll>(n, 0), vector<ll>(n, 0)};
        for (int j = 0; j < n; ++j){
            buff[0][j] = points[0][j];
        }
        for (int i = 1; i < m; ++i){
            ll opt = buff[0][0];
            buff[1][0] = opt + points[i][0];
            for (int j = 1; j < n; ++j){
                opt = max(opt - 1, buff[0][j]);
                buff[1][j] = opt + points[i][j];
            }
            opt = 0;
            for (int j = n - 2; j >= 0; --j){
                opt = max(opt, buff[0][j + 1]) - 1;
                buff[1][j] = max(buff[1][j], opt + points[i][j]);
            }
            swap(buff[0], buff[1]);
        }
        return *ranges::max_element(buff[0]);
    }
};
